### ISSUE TITLE
Playback hangs or audio drops

### ISSUE THREAD

**Issue by cookieisland** [rocky4546/tvheadend-locast#3]
I am getting hangs or audio drops.  It seems to happen when a free account refresh occurs.  It doesn't happen for every refresh. I am guessing I get a hang or audio dropout every hour or so. When I look at the log from tvheadend-locast it looks like the refresh occurs but one or more '.ts' urls are skipped (BTW: when the refresh works no '.ts' files are skipped). When the hang or audio drop happens I quickly stop/start the playback and I can see that in the log as well. I am using a free Locast account and my hardware/sw config is CorELEC v19 (Matrix v19) on an Odroid N2, and TVheadend and TVheadend-locast via docker containers in an Ubuntu/Intel vm.  I am using stream_type = internalproxy.  Here is a piece of my log when it happened.

2021-03-20 18:54:23,825-INFO: Serving https://sfo.locastnet.org/proxy/sfo/kntv_src269381952.ts (6.0s) in, 2.55s
2021-03-20 18:54:33,016-INFO: Serving https://sfo.locastnet.org/proxy/sfo/kntv_src269381953.ts (6.0s) in, 2.60s
2021-03-20 18:54:41,726-INFO: Refresh time expired. Refresh rate is 800 seconds
Getting station info for 1316...
Determining best video stream for 1316...
Found 3 Playlists
1316 will use 1280x720 resolution at 2700000bps
2021-03-20 18:54:42,974-INFO: Serving https://sfo.locastnet.org/proxy/sfo/kntv_src269381955.ts (6.0s) in, 2.75s
2021-03-20 18:54:52,173-INFO: Serving https://sfo.locastnet.org/proxy/sfo/kntv_src269381956.ts (6.0s) in, 2.75s
2021-03-20 18:54:55,272-INFO: Serving https://sfo.locastnet.org/proxy/sfo/kntv_src269381957.ts (6.0s) in, 2.77s
Getting station info for 1316...
Determining best video stream for 1316...
Found 3 Playlists
1316 will use 1280x720 resolution at 2700000bps
2021-03-20 18:55:04,491-INFO: Serving https://sfo.locastnet.org/proxy/sfo/kntv_src269381958.ts (6.0s) in, 2.77s
2021-03-20 18:55:04,492-INFO: 2. Connection dropped by end device
2021-03-20 18:55:04,493-INFO: Locast Connection Closed
2021-03-20 18:55:04,576-INFO: Serving https://sfo.locastnet.org/proxy/sfo/kntv_src269381958.ts (6.0s) in, 2.68s
2021-03-20 18:55:07,608-INFO: Serving https://sfo.locastnet.org/proxy/sfo/kntv_src269381959.ts (6.0s) in, 2.75s
2021-03-20 18:55:16,815-INFO: Serving https://sfo.locastnet.org/proxy/sfo/kntv_src269381960.ts (6.0s) in, 2.77s
2021-03-20 18:55:20,005-INFO: Serving https://sfo.locastnet.org/proxy/sfo/kntv_src269381961.ts (6.0s) in, 2.70s
2021-03-20 18:55:29,200-INFO: Serving https://sfo.locastnet.org/proxy/sfo/kntv_src269381962.ts (6.0s) in, 2.72s

**Comment by rocky4546** [comment]
Thanks for the log and the info.  This issue makes sense.  On refresh, locast will reset and provide the latest 6 second stream.  If the last stream was near 6 seconds ago, then it would skip a stream.  Looking at the log, the last stream was at 54:33 with the refresh at 54:41 (8 seconds later), so it is confirmed.  I will make a change to not check for a timeout unless an update occurs.  This should stop the skip.  I just finished the updates to this release and am working on getting the release out today.  Hopefully, I checked it out so no additional bugs are present.  If you want to try out the bug fix, here it is.

remove lines from around line 556
```
                if self.is_time_to_refresh():
                    stream_uri = self.locast.get_station_stream_uri(sid)
                    self.last_refresh = time.time()
```
modify lines:
```
                 if added == 0 and duration > 0:
                    time.sleep(duration)
```
to
```
                if added == 0 and duration > 0:
                    time.sleep(duration)
                else:
                    if self.is_time_to_refresh():
                        stream_uri = self.locast.get_station_stream_uri(sid)
                        self.last_refresh = time.time()
```

This should only do a refresh when a new item is added by locast to the stream.  If that is not enough, then also change

`                        target_diff = 0.5 * duration`
to
`                        target_diff = 0.3 * duration`

let me know if that works.

**Comment by cookieisland** [comment]
I was just now doing some testing with full debug on to see what was happening...   I suspected a timing issue was happening and was going to diagnose and try to fix but since you diagnosed it with the info I sent that is great and I will test your changes.

**Comment by rocky4546** [comment]
One more change to fully fixed it.  For testing, I set the refresh rate to 120, so it happens more often.  The change I recommend is around line 589.
change
                    time.sleep(duration)
to
                    time.sleep(duration*0.3)

What I look for is ... after the reset, does it only remove items from the list or does it also add.  Best result is to only remove items.  Then after a short timeout, it adds one back in. You will need DEBUG on to see the queue changes.

Have a few cleanup items for this bug and will get back to the release.

**Comment by cookieisland** [comment]
I was running your first change for a couple of hours and did not run into any problems (I had set refresh to 180 secs). I will incorporate your 2nd change as well...

**Comment by cookieisland** [comment]
With your first change I did see some debug messages like this after a refresh:
2021-03-23 16:53:02,538-DEBUG: Added 1 new segments, removed 10

I am making the duration change now and will continue to test.

**Comment by cookieisland** [comment]
The testing I did with the "time.sleep(duration*0.3)" change would still periodically add an item without a timeout.

I am now running the latest version 0.7.5 and so far so good....  no adding of items until a subsequent timeout - at least so far anyway.


BTW: I noticed a couple of other things

1) The config.ini file gets overwritten so if I have a comment to myself in the file it gets removed.  This is probably not a huge deal and I just need to use the html settings page vs editing the config.ini file.  The reason I had some comments in the config.ini file was some notes to myself and also to make it easy to turn on/off options. eg I had the following in the file so I could easily switch to a 2 minute refresh for testing purposes.
[freeaccount]
#refresh_rate = 120

 
2) When doing a new install with no previous cache it takes a fairly long time to get the EPG populated which may lead a user to beleive they did something wrong. I think what happens is when cache is empty epg2xml.py creates a dummy EPG file and then immediately calls generate_epg_file.  generate_epg_file sees that the dummy file is not yet 4 hours old (min_refresh_rate) so it quits.  The problem I see is that new user will install tvheadend-locast and then tries to add the EPG to tvheadend.  They will not get an EPG for 4 hours and they may think they did something wrong or something is broken.....  I noticed this with 0.7.4 and started searching thru the code to figure out what was wrong.  After I looked at the code I found I could just open a shell in the docker container and "touch -d yesterday /app/cache/xxx_epg.xml" and then things would work.  Maybe when <ip>:6077/xmltv is requested a check can be made to see if the xxx_epg.xml file is a dummy and if so, generate a new one???

**Comment by rocky4546** [comment]
Thanks for upgrading.  

As for #1.  Yes, ConfigParser does that. It looks like there is a way to keep comments by making the configparser think that a comment is a name/value without a value.  I will look into it.  Also, recommend trying out the new website.  Just go to :6077/ The config editor (which includes updates to the website themes) and the list of json/xml links available work fine. I checked it out on Firefox, then Chrome and Edge.  The config display is based on the config.json file in the tvheadend folder.

I will check on #2.  I did put some code into that area to make it more pro-active for epg updates.  It is possible when initializing, it may have some issue.  Will get back to you on that.

Just to let you know, the next upgrade involves modularization of Locast into a provider addon and incorporating PlutoTV as another provider addon.

**Comment by cookieisland** [comment]
I added a PR trying to address #2 - I simply adjusted the timestamp on the dummy epg file to short circuit the timer...

PlutoTV update sounds interesting.  Tying into tvheadend works quite well for Locast so hopefully PlutoTV also ties in well too.

Thanks again for this great work. Locast is running very well for me now.

**Comment by rocky4546** [comment]
Found a minor bug in the website config editor.  If a value is null, it will throw a javascript error and all remaining values are empty.  I have a fix in the javascript file.  As for the ConfigParser, I tried a few recommendations and determined that the class has bugs in this area and basically does not work to allow comments.  The alternative is to not use the python default class and to use ConfigObj, which is a github product.  Not really wanting to go there at this point, so probably just need to deal with the limitation.  I would hope a future release of the parser would fix the issues since most developers are having to use ConfigObj.

Since this issue is resolve, will close it.
